master/buildbot/db/model.py

```diff
@@ -988,35 +992,51 @@ class Model(base.DBConnectorComponent):
     # Migration support
     # -----------------
 
-    # this is a bit more complicated than might be expected because the first
-    # seven database versions were once implemented using a homespun migration
-    # system, and we need to support upgrading masters from that system.  The
-    # old system used a 'version' table, where SQLAlchemy-Migrate uses
-    # 'migrate_version'
+    # Buildbot has historically used 3 database migration systems:
+    #  - homegrown system that used "version" table to track versions
+    #  - SQLAlchemy-migrate that used "migrate_version" table to track versions
+    #  - alembic that uses "alembic_version" table to track versions (current)
+    # We need to detect each case and tell the user how to upgrade.
+
+    config_path = util.sibpath(__file__, "migrations/alembic.ini")
 
-    repo_path = util.sibpath(__file__, "migrate")
+    def table_exists(self, conn, table):
+        try:
+            r = conn.execute(f"select * from {table} limit 1")
```

Comment:
I know this is an internal function but [it would be safer](https://xkcd.com/327/) to use:

```python
r = conn.execute("select * from ? limit 1", table)
```

Reply:
This doesn't work unfortunately. I've tried a couple of different ways to do this and sqlalchemy seems to be not able to add the parameter to the query.